### PR TITLE
add mypy annotations

### PR DESCRIPTION
**Pull request by graingert**

### CODE REVIEW

File: src/incremental/__init__.py [hunk 1]
```diff
@@ -11,105 +11,56 @@
...
-            return 0
-        else:
-            return 1
+    def _cmp(a, b):  # type: (Any, Any) -> int
+        return 0
 
 
-def _comparable(klass):
-    """
-    Class decorator that ensures support for the special C{__cmp__} method.
```
Comment (graingert) [inline src/incremental/__init__.py:45]: this decorator is broken on mypy - and the type annotations are slightly different,

eg Version is only comparable with other Versions

File: src/incremental/update.py [hunk 2]
```diff
@@ -117,30 +156,39 @@ def _run(
...
         if len(release) == 1:
-            release.append(0)
-        if len(release) == 2:
-            release.append(0)
+            (major,) = release
+        elif len(release) == 2:
+            (
+                major,
+                minor,
+            ) = release
```
Comment (graingert) [inline src/incremental/update.py:172]:
```suggestion
            major, minor = release
```

File: src/incremental/update.py [hunk 3]
```diff
@@ -6,9 +6,51 @@
...
+        def write(self, v):  # type: (bytes) -> object
+            pass
+
+        def __enter__(self):  # type: () -> _ReadableWritable
+            pass
+
+        def __exit__(self, *args, **kwargs):  # type: (object, object) -> Optional[bool]
+            pass
+
+    class FilePath(object):
```
Comment (graingert) [inline src/incremental/update.py:29]: FilePath is type annotated in twisted, but it's decorated with `twisted.python.comparable` which strips the annotations

File: src/incremental/update.py [hunk 4]
```diff
@@ -117,30 +156,36 @@ def _run(
...
+        micro = 0
         if len(release) == 1:
-            release.append(0)
-        if len(release) == 2:
-            release.append(0)
+            (major,) = release
+        elif len(release) == 2:
+            major, minor = release
+        else:
+            major, minor, micro = release
```
Comment (altendky) [inline src/incremental/update.py:171]: If only Python 3...  oh well.

```python
if len(release) > 3:
    raise Exception("ack!")

major, minor, micro, [_] = [*release, 0, 0, 0]
```

**Review by altendky — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 5]
```diff
@@ -1,3 +1,10 @@
...
-requires = [ "setuptools >= 44.1.1", "wheel >= 0.36.2"]
+requires = [
+    "setuptools >= 44.1.1",
+    "wheel >= 0.36.2",
+    "typing >= 3.7.4.3; python_version < '3.5'",
+]
 build-backend = "setuptools.build_meta"
+
+[tool.black]
+target-version = ['py27', 'py36', 'py37', 'py38', 'py38']
```
[the remark this answers is not in the document]
Comment (graingert) [reply pyproject.toml:10]: black doesn't support py39 as a target

File: src/incremental/__init__.py [hunk 6]
```diff
@@ -121,8 +72,22 @@ def __cmp__(self, other):
             return 0
         return 1
 
+    if sys.version_info >= (3, 0):
```
[the remark this answers is not in the document]
Comment (graingert) [reply src/incremental/__init__.py:75]: I don't know - I was planning on dropping py2.7 support and not thinking about it

File: src/incremental/update.py [hunk 7]
```diff
@@ -78,12 +120,9 @@ def _run(
...
 
     if type(package) != str:
-        package = package.encode("utf8")
+        package = package.encode("utf8")  # type: ignore[assignment]
 
-    if not path:
-        path = _findPath(_getcwd(), package)
-    else:
-        path = FilePath(path)
+    _path = FilePath(path) if path else _findPath(_getcwd(), package)
```
[the remark this answers is not in the document]
Comment (graingert) [reply src/incremental/update.py:125]: mypy doesn't let you redefine an `Optional[str]` as an `FilePath`

File: src/incremental/update.py (hunk 3 above)
Comment (graingert) [inline src/incremental/update.py:29]:
```suggestion
    # FilePath is missing type annotations
    # https://twistedmatrix.com/trac/ticket/10148
    class FilePath(object):
```

File: pyproject.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (graingert) [reply pyproject.toml:10]:
```suggestion
target-version = ['py27', 'py36', 'py37', 'py38']
```

**Review by altendky — APPROVED** [review]
Approved pending CI success.

**Pull request merged**